packages/discord.js/src/structures/interfaces/Collector.js

```diff
@@ -171,12 +171,59 @@ class Collector extends EventEmitter {
 
       const cleanup = () => {
         this.removeListener('collect', onCollect);
+        this.removeListener('dispose', onDispose);
+        this.removeListener('ignore', onIgnore);
         this.removeListener('end', onEnd);
       };
 
-      const onCollect = item => {
+      const onCollect = (...items) => {
         cleanup();
-        resolve(item);
+        resolve(['collecting', ...items]);
+      };
+
+      const onDispose = (...items) => {
+        cleanup();
+        resolve(['disposing', ...items]);
+      };
+
+      const onIgnore = (...items) => {
+        cleanup();
+        resolve(['ignoring', ...items]);
+      };
+
+      const onEnd = () => {
+        cleanup();
+        reject(this.collected);
+      };
+
+      this.on('collect', onCollect);
+      this.on('dispose', onDispose);
+      this.on('ignore', onIgnore);
+      this.on('end', onEnd);
+    });
+  }
+
+  /**
+   * Returns a promise that resolves with the next collected elements;
+   * rejects with collected elements if the collector finishes without receiving a next element
+   * @type {Promise}
+   * @readonly
+   */
+  get nextCollecting() {
```

Comment:
I feel like we shouldn't have async getters. What about `waitForCollect()`? Same for the other methods.

I also think that it'd be useful to accept an `AbortSignal` so we can cancel and unregister the listeners earlier.